examples/hdf5.py

```diff
@@ -4,99 +4,199 @@
-        ds = int((stop-start) / self.limit) + 1
-        
-        if ds == 1:
-            # Small enough to display with no intervention.
-            visible = self.hdf5[start:stop]
-            scale = 1
-        else:
-            # Here convert data into a down-sampled array suitable for visualizing.
-            # Must do this piecewise to limit memory usage.        
-            samples = 1 + ((stop-start) // ds)
-            visible = np.zeros(samples*2, dtype=self.hdf5.dtype)
-            sourcePtr = start
-            targetPtr = 0
-            
-            # read data in chunks of ~1M samples
-            chunkSize = (1000000//ds) * ds
-            while sourcePtr < stop-1: 
-                chunk = self.hdf5[sourcePtr:min(stop,sourcePtr+chunkSize)]
-                sourcePtr += len(chunk)
-                
-                # reshape chunk to be integral multiple of ds
-                chunk = chunk[:(len(chunk)//ds) * ds].reshape(len(chunk)//ds, ds)
-                
-                # compute max and min
-                chunkMax = chunk.max(axis=1)
-                chunkMin = chunk.min(axis=1)
-                
-                # interleave min and max into plot data to preserve envelope shape
-                visible[targetPtr:targetPtr+chunk.shape[0]*2:2] = chunkMin
-                visible[1+targetPtr:1+targetPtr+chunk.shape[0]*2:2] = chunkMax
-                targetPtr += chunk.shape[0]*2
-            
-            visible = visible[:targetPtr]
-            scale = ds * 0.5
-            
-        self.setData(visible) # update the plot
-        self.setPos(start, 0) # shift to match starting index
-        self.resetTransform()
-        self.scale(scale, 1)  # scale to match downsampling
 
-        
+        x1, x2 = vb.viewRange()[0]
+        if not lazy or not self.sampler.isRangeValid(x1, x2):
+            x, y = self.sampler[x1:x2]
+            self.setData(x, y)  # update the plot
+
+
+class DataSampler:
+    def __init__(self, x, y, sampleLimit=2500, cacheLimit=10000000, padding=0.3, eager=True):
+        if x is not None and len(x) != len(y):
+            raise ValueError
+
+        self.x = x
+        self.y = y
+        self.sampleLimit = sampleLimit  # at most 4 times this number of samples will be plotted
+        self.cacheLimit = cacheLimit  # at most 8 times this number of samples will be cached
+        self.cacheSize = 0
+        self.padding = padding
+        self._minCacheLevel = max(2, (len(self) // self.cacheLimit).bit_length())  # n < 2 (ds < 4) has no effect
+        self._maxCacheLevel = (len(self) // self.sampleLimit).bit_length()
+        self._last = None
+        if eager and (self._maxCacheLevel >= self._minCacheLevel):
+            # spawn a thread to eagerly downsample data up to self._maxCacheLevel so
+            # it will be ready for when the user zooms out
+            self._t = threading.Thread(target=lambda: self._loadLevel(self._maxCacheLevel))
```

Comment:
why do you have this `lambda`?
It's a shared mem thread, just read the `._maxCacheLevel` from the target if you're trying to read it "lazily".

Reply:
No good reason. It just evolved that way to be taking as an argument what eventually turned into an attribute.

Reply:
Even still, it's not really lazy since you start the thread right away.

Reply:
We don't use it anywhere else but `QtConcurrent` and `QFuture` might be worthwhile here; although looks like the threading code needed here is pretty minimal.

Reply:
Oh, sorry, I didn't think long enough about your question before answering. `_loadLevel` needs to take the argument explicitly because it's recursively defined.

Reply:
I still don't grok.

- the bare `lambda` is odd / unrequired.
- you could still do this with a function and not a method 

Maybe you want a `functools.partial` or just use the `args` to [`threading.Thread`](https://docs.python.org/3/library/threading.html#threading.Thread)?